f5_openstack_agent/lbaasv2/drivers/bigip/disconnected_service.py

```diff
@@ -0,0 +1,149 @@
+                            if (provisioning_status !=
+                                    plugin_const.PENDING_DELETE):
+                                d.plugin_rpc.update_loadbalancer_status(id)
+                    else:
+                        self.start_time[id] = timeutils.utcnow()
+                        d.plugin_rpc.update_loadbalancer_status(
+                            id, provisioning_status, lb_const.OFFLINE)
+                    continue
+                # service is connected in neutron, move all listeners for this
+                # loadbalancer onto a real network
+                for listener in listeners:
+                    service['listener'] = listener
+                    virtual = d.service_adapter.get_virtual_name(service)
+                    bigips = d.get_all_bigips()
+                    if not d.disconnected_service.is_virtual_connected(
+                            virtual, bigips):
+                        LOG.debug("connecting %s to a real network" %
+                                  virtual['name'])
+                        d.lbaas_builder.listener_builder.update_listener(
+                            service, bigips)
+                d.plugin_rpc.update_loadbalancer_status(
+                    id, provisioning_status, lb_const.ONLINE)
+                self.start_time.pop(id, None)
+
+    def polling_thread(self):
+        while True:
+            # Split out the actual scanning tech to accommodate migration
+            # greeenthread to oslo periodic_task.
+            self.scan()
+            greenthread.sleep(
+                self.driver.conf.f5_network_segment_polling_interval)
+
+
+class DisconnectedService(object):
+    network_name = 'disconnected_network'
+
+    def __init__(self):
+        self.network_name = DisconnectedService.network_name
+        self.network_helper = NetworkHelper()
+
+    # The following method presumes that the plugin driver is aware that we're
+    # running in hierarchical mode or not and sets segmentation_id correctly.
+    def is_service_connected(self, service):
+        networks = service['networks']
+        network_id = service['loadbalancer']['network_id']
+        segmentation_id = networks[network_id]['provider:segmentation_id']
+        return (segmentation_id)
+
+    def is_virtual_connected(self, virtual, bigips):
+        # check if virtual_server is connected on any of our bigips
+        connected = True
+        for bigip in bigips:
+            vs = bigip.ltm.virtuals.virtual
+            if vs.exists(name=virtual['name'], partition=virtual['partition']):
+                vs.load(name=virtual['name'], partition=virtual['partition'])
+                if (getattr(vs, 'vlansDisabled', False) or
+                        not getattr(vs, 'vlansEnabled', True)):
+                    # accommodate quirk of how big-ip returns virtual server
+                    # if vlans are disabled OR vlans are not enabled, then
+                    # we're connected
```

Comment:
I was thinking about this code and the BIG-IP's behavior and I think it is safe to say that you will always only ever have one or the other parameters in the JSON that is returned by BIG-IP and the one that is returned will always be the one that is set to `True`. I am fine how you have it, but you can probably just do the first one half and be fine.

```python
if (vs.get('vlansDisabled', False):
    ...
```